llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -9093,19 +9112,38 @@ addUsersInExitBlocks(VPlan &Plan,
       if (Op->isLiveIn())
         continue;
 
-      // Currently only live-ins can be used by exit values from blocks not
-      // exiting via the vector latch through to the middle block.
-      if (ExitIRI->getParent()->getSinglePredecessor() != MiddleVPBB)
-        return false;
-
       LLVMContext &Ctx = ExitIRI->getInstruction().getContext();
-      VPValue *Ext = B.createNaryOp(VPInstruction::ExtractFromEnd,
-                                    {Op, Plan.getOrAddLiveIn(ConstantInt::get(
-                                             IntegerType::get(Ctx, 32), 1))});
+      VPValue *Ext;
+      VPBasicBlock *PredVPBB =
+          cast<VPBasicBlock>(ExitIRI->getParent()->getPredecessors()[Idx]);
+      if (PredVPBB != MiddleVPBB) {
+        assert(ExitIRI->getParent()->getNumPredecessors() <= 2);
+
+        // Cache the early exit mask
```

Comment:
```suggestion
        // Lookup and cache the early exit mask.
```

Caching is done to avoid repeated lookups, i.e. compile time perf opt?

Reply:
Yeah that's right. If we don't store the mask in the VPRegionBlock then we have to calculate it manually and I was hoping to avoid constant recalculation of the same value. I could also just calculate this prior to the loop, but then the work is unnecessary if there aren't any early exits.

Reply:
Done